src/Npgsql/NpgsqlConnector.Auth.cs

```diff
@@ -73,7 +73,7 @@ async Task AuthenticateSASL(List<string> mechanisms, bool async)
                                           "Mechanisms received from server: " + string.Join(", ", mechanisms));
             var mechanism = "SCRAM-SHA-256";
 
-            var passwd = GetPassword() ??
+            var passwd = GetPassword(Settings.Username) ??
```

Comment:
As noted above, we have a `GetUsername()` fallback which attempts to find out the username from the OS, so we shouldn't access Settings.Username directly.

Reply:
`GetUsername()` throws if one can't be found, nowhere else in `AuthenticateSASL()` is a username referenced. 
I'm not familiar with this code so i'm not 100% sure this isn't going to break if i change this to 
```cs
var passwd = GetPassword(GetUsername()) ??
    throw new NpgsqlException($"No password has been provided but the backend requires one (in SASL/{mechanism})");
```

Reply:
i did this in https://github.com/npgsql/npgsql/pull/2502/commits/3757753643e24df9ed4b7291385ff23f61ec9e0b and all tests are still passing.

However i deleted the body of `AuthenticateSASL()` and replaced it with an exception and all tests still pass, so this isn't a confidence booster that i haven't broken this authentication scheme.

Reply:
I think AuthenticateSASL should simply accept a username parameter like the cleartext/MD5 - there's no need to call GetUsername again. For our purposes SASL is just another mechanism for encoding the password, stronger than MD5.

Note that we do have GSS, which really is password-less authentication - this is why it doesn't need a username or password.